psi4/src/psi4/libmints/eri.cc

```diff
@@ -179,10 +179,31 @@ Libint2ERI::Libint2ERI(const IntegralFactory *integral, double screening_thresho
                              std::max(basis3()->max_nprimitive(), basis4()->max_nprimitive()));
     const auto max_precision = std::numeric_limits<double>::epsilon() * 1e-13;
     const auto engine_precision = screening_threshold == 0.0 ? max_precision : std::numeric_limits<double>::epsilon() * std::pow((double)max_nprim, -4.0);
+
+    bool dummy1 = basis1()->l2_shell(0) == libint2::Shell::unit();
+    bool dummy2 = basis2()->l2_shell(0) == libint2::Shell::unit();
+    bool dummy3 = basis3()->l2_shell(0) == libint2::Shell::unit();
+    bool dummy4 = basis4()->l2_shell(0) == libint2::Shell::unit();
+
+     if (!dummy1 && !dummy2 && !dummy3 && !dummy4) {
+        braket_ = libint2::BraKet::xx_xx;
+    } else if (!dummy1 && dummy2 && !dummy3 && !dummy4) {
+        braket_ = libint2::BraKet::xs_xx;
+    } else if (!dummy1 && !dummy2 && !dummy3 && dummy4) {
+        braket_ = libint2::BraKet::xx_xs;
+    } else if (!dummy1 && dummy2 && !dummy3 && dummy4) {
+        braket_ = libint2::BraKet::xs_xs;
+    } else {
+        throw PSIEXCEPTION("Bad BraKet type in Libint2TwoElectronInt");
+    }
+
     for (int der = 0; der <= deriv; ++der) {
-        engines_.emplace_back(libint2::Operator::coulomb, max_nprim, max_am, der, engine_precision);
+        // set braket upon engine construction so particular LIBINT2_MAX_AM_eri[|2|3] limit governs validity
+        engines_.emplace_back(libint2::Operator::coulomb, max_nprim, max_am, der, engine_precision, libint2::operator_traits<libint2::Operator::coulomb>::default_params(), braket_);
     }
-    schwarz_engine_ = libint2::Engine(libint2::Operator::coulomb, max_nprim, max_am, 0, max_precision);
+    // set max_am for primary basis to be sieved, not all basis1234
+    max_am = std::max(std::min(basis1()->max_am(), basis2()->max_am()), std::min(basis3()->max_am(), basis4()->max_am()));
```

Comment:
I think this can be simplified to `max_am = bra_same_ ? basis1()->max_am() : ket_same_ ? basis3()->max_am() : 0;`. Those variables are defined upstream in twobody.cc; essentially if we find a matching pair the bra, that's what will be used for the sieve, else if we find a matching pair in the ket, the ket basis will be used.  If neither bra nor ket match, the integral object doesn't do any sieving - there's some specialized sieve code in the DF JK object than handles things like (A0|B0) for aux integrals (where 0 is the empty basis).

Reply:
@andysim I'm not sure that's simpler :laughing:

Reply:
Oops, I was misreading.  @loriab's solution gives the exact same results and, as you point out, is probably easier to read.  My concern was that it would end up generating high angular momentum requirements for DF integrals but it does indeed yield 0.  Probably best to leave it as-is